File: src/lib/api/ipset.ts
/**
 * IP Set management API
 * 
 * Functions for managing IP address lists (ipset) for DPI bypass.
 * Supports downloading from remote sources and auto-updates.
 */

import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Types
// ============================================================================

/**
 * Source configuration for an IP set
 */
export interface IpsetSource {
    /** Source ID */
    id: string;
    /** Human-readable name */
    name: string;
    /** Remote URL to download from */
    url: string;
    /** Optional description */
    description: string | null;
    /** Whether this source is enabled */
    enabled: boolean;
    /** Priority (lower = higher priority) */
    priority: number;
}

/**
 * Information about an ipset for UI display
 */
export interface IpsetInfo {
    /** Ipset ID (filename without extension) */
    id: string;
    /** Human-readable name */
    name: string;
    /** Last update timestamp (ISO 8601) */
    last_updated: string | null;
    /** File size in bytes */
    size: number | null;
    /** Total number of IP entries */
    ip_count: number | null;
    /** Number of IPv4 entries */
    ipv4_count: number | null;
    /** Number of IPv6 entries */
    ipv6_count: number | null;
    /** Number of CIDR entries */
    cidr_count: number | null;
    /** Whether update is available */
    update_available: boolean;
    /** Source URL (if configured) */
    source_url: string | null;
    /** Whether auto-update is enabled */
    auto_update_enabled: boolean;
}

/**
 * Result of updating ipset
 */
export interface IpsetUpdateResult {
    /** Whether update was successful */
    success: boolean;
    /** Number of IP entries after update */
    ip_count: number;
    /** Number of IPv4 entries */
    ipv4_count: number;
    /** Number of IPv6 entries */
    ipv6_count: number;
    /** Number of CIDR entries */
    cidr_count: number;
    /** Source URL used */
    source_url: string;
    /** Error message if failed */
    error: string | null;
    /** Timestamp of update (ISO 8601) */
    timestamp: string;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get information about the current ipset.
 * 
 * Returns metadata including IP count, last update time, and source URL.
 * 
 * @returns Current ipset info
 */
export async function getIpsetInfo(): Promise<IpsetInfo> {
    return invoke<IpsetInfo>('get_ipset_info');
}

/**
 * Update ipset from a specific source URL.
 * 
 * Downloads IP addresses from the given URL and updates ipset-all.txt.
 * Validates content before saving (only IPv4/IPv6 and CIDR allowed).
 * 
 * @param sourceUrl - URL to download ipset from
 * @returns Update result with IP counts
 */
export async function updateIpset(sourceUrl: string): Promise<IpsetUpdateResult> {
    return invoke<IpsetUpdateResult>('update_ipset', { sourceUrl });
}

/**
 * Update ipset from configured sources.
 * 
 * Tries each configured source in priority order until one succeeds.
 * Primary source is zapret-discord-youtube GitHub repository.
 * 
 * @returns Update result with IP counts
 */
export async function updateIpsetFromSources(): Promise<IpsetUpdateResult> {
    return invoke<IpsetUpdateResult>('update_ipset_from_sources');
}

/**
 * Set ipset auto-update enabled/disabled.
 * 
 * When enabled, ipset will be automatically updated once per day.
 * 
 * @param enabled - Whether auto-update should be enabled
 */
export async function setIpsetAutoUpdate(enabled: boolean): Promise<void> {
    return invoke<void>('set_ipset_auto_update', { enabled });
}

/**
 * Get ipset sources configuration.
 * 
 * Returns the list of configured ipset sources with their URLs and priorities.
 * 
 * @returns List of configured sources
 */
export async function getIpsetSources(): Promise<IpsetSource[]> {
    return invoke<IpsetSource[]>('get_ipset_sources');
}

/**
 * Restore ipset from backup.
 * 
 * Restores the previous version of ipset-all.txt from backup.
 * Useful if an update caused issues.
 */
export async function restoreIpsetBackup(): Promise<void> {
    return invoke<void>('restore_ipset_backup');
}

File: src-tauri/src/commands/ipset.rs
//! IP Set management commands
//!
//! Commands for managing IP address lists (ipset) for DPI bypass.
//! Supports downloading from remote sources, auto-updates, and mode switching.
//!
//! ## Modes
//! - `any`: Allow all traffic (ipset filtering disabled)
//! - `none`: Block all traffic that would match ipset rules
//! - `loaded`: Use loaded ipset for filtering

use tracing::info;

use crate::commands::validation::validate_public_url;
use crate::core::errors::{IsolateError, TypedResultExt};
use crate::core::ipset_manager::{
    self, IpsetMode, IpsetStats,
};
use crate::core::ipset_updater::{
    self, IpsetInfo, IpsetUpdateResult, IpsetUpdater, IpsetValidationResult,
};

/// Get information about the current ipset
/// 
/// Returns metadata including IP count, last update time, and source URL.
#[tauri::command]
pub async fn get_ipset_info() -> Result<IpsetInfo, IsolateError> {
    info!("Getting ipset info");

    ipset_updater::get_ipset_info()
        .await
        .io_context("Failed to get ipset info")
}

/// Update ipset from a specific source URL
/// 
/// Downloads IP addresses from the given URL and updates ipset-all.txt.
/// Validates content before saving (only IPv4/IPv6 and CIDR allowed).
/// 
/// Security: URL is validated to prevent SSRF attacks (blocks localhost, private IPs).
#[tauri::command]
pub async fn update_ipset(source_url: String) -> Result<IpsetUpdateResult, IsolateError> {
    // SSRF protection: validate URL points to public address only
    let validated_url = validate_public_url(&source_url)?;
    
    info!(url = %validated_url, "Updating ipset from URL");

    ipset_updater::update_ipset(validated_url.as_str())
        .await
        .network_context("Failed to update ipset")
}

/// Update ipset from configured sources
/// 
/// Tries each configured source in priority order until one succeeds.
/// Sources are configured in configs/ipset_sources.yaml.
#[tauri::command]
pub async fn update_ipset_from_sources() -> Result<IpsetUpdateResult, IsolateError> {
    info!("Updating ipset from configured sources");

    ipset_updater::update_ipset_from_sources()
        .await
        .network_context("Failed to update ipset from sources")
}

/// Set ipset auto-update enabled/disabled
/// 
/// When enabled, ipset will be automatically updated once per day.
#[tauri::command]
pub async fn set_ipset_auto_update(enabled: bool) -> Result<(), IsolateError> {
    info!(enabled = enabled, "Setting ipset auto-update");

    ipset_updater::set_ipset_auto_update(enabled)
        .await
        .io_context("Failed to set ipset auto-update")
}

/// Get ipset sources configuration
/// 
/// Returns the list of configured ipset sources with their URLs and priorities.
#[tauri::command]
pub async fn get_ipset_sources() -> Result<Vec<ipset_updater::IpsetSource>, IsolateError> {
    info!("Getting ipset sources");

    let updater = IpsetUpdater::new()
        .await
        .io_context("Failed to create ipset updater")?;
    
    Ok(updater.get_sources().to_vec())
}

/// Restore ipset from backup
/// 
/// Restores the previous version of ipset-all.txt from backup.
#[tauri::command]
pub async fn restore_ipset_backup() -> Result<(), IsolateError> {
    info!("Restoring ipset from backup");

    let updater = IpsetUpdater::new()
        .await
        .io_context("Failed to create ipset updater")?;
    
    updater.restore_from_backup()
        .await
        .io_context("Failed to restore ipset backup")
}

// ============================================================================
// Ipset Manager Commands (Mode Management)
// ============================================================================

/// Get ipset statistics including mode, IP counts, and update info
/// 
/// Returns comprehensive statistics about the current ipset state.
#[tauri::command]
pub async fn get_ipset_stats() -> Result<IpsetStats, IsolateError> {
    info!("Getting ipset stats");

    ipset_manager::get_ipset_stats()
        .await
        .io_context("Failed to get ipset stats")
}

/// Get current ipset mode
/// 
/// Returns the current mode: "any", "none", or "loaded".
#[tauri::command]
pub async fn get_ipset_mode() -> Result<IpsetMode, IsolateError> {
    info!("Getting ipset mode");

    ipset_manager::get_current_mode()
        .await
        .io_context("Failed to get ipset mode")
}

/// Set ipset mode
/// 
/// Changes the ipset operation mode:
/// - `any`: Allow all traffic (ipset filtering disabled)
/// - `none`: Block all traffic that would match ipset rules  
/// - `loaded`: Use loaded ipset for filtering
/// 
/// Note: Switching to "loaded" mode requires an ipset file to exist.
#[tauri::command]
pub async fn set_ipset_mode(mode: String) -> Result<(), IsolateError> {
    info!(mode = %mode, "Setting ipset mode");

    let mode: IpsetMode = mode.parse()
        .map_err(|e: IsolateError| e)?;

    ipset_manager::set_current_mode(mode)
        .await
        .io_context("Failed to set ipset mode")
}

/// Load ipset from a local file
/// 
/// Loads and validates an ipset file from the specified path.
/// The file is copied to the default location for use.
#[tauri::command]
pub async fn load_ipset_from_file(path: String) -> Result<IpsetValidationResult, IsolateError> {
    info!(path = %path, "Loading ipset from file");

    ipset_manager::load_ipset_from_file(&path)
        .await
        .io_context("Failed to load ipset from file")
}

/// Update ipset from a specific URL (via manager)
/// 
/// Downloads and updates ipset from the given URL.
/// This is an alternative to `update_ipset` that goes through the manager.
#[tauri::command]
pub async fn update_ipset_from_url(url: String) -> Result<IpsetUpdateResult, IsolateError> {
    // SSRF protection: validate URL points to public address only
    let validated_url = validate_public_url(&url)?;
    
    info!(url = %validated_url, "Updating ipset from URL via manager");

    ipset_manager::update_ipset_from_url(validated_url.as_str())
        .await
        .network_context("Failed to update ipset from URL")
}
